From 717c5ffe46e038ffc89e8651ad009e85f71c3bda Mon Sep 17 00:00:00 2001
From: Jeff Kletsky <git-commits@allycomm.com>
Date: Sat, 28 May 2022 09:39:01 -0700
Subject: [PATCH] ipq40xx: ea8300, mr8300: DSA: Adjust MAC addresses

At present, all four Ethernet phys associated with the lan0-3 ports
are assigned the same MAC address. This causes MAC collisions that
become problematic with more complex network configurations.

This patch assigns the "label MAC" to "wan" and the next sequential
to "lan1". The MAC assigned to lan1 was used by the OEM firmware.
lan2-4 are assigned locally administered MAC addresses, sequential
from lan1 with its LA bit set. "eth0" is assigned the next in sequence,
pushing the generated MACs above the range already in use.

label_mac is also defined and now available in board.json

NB: If MAC assignments are already present in /etc/config/network
    they take precedence. These assignments may already be present
    from an eariler boot sequence that generated the local config.

Tested-on: ea8300

Signed-off-by: Jeff Kletsky <git-commits@allycomm.com>
---
 target/linux/ipq40xx/base-files/etc/board.d/02_network     | 1 +
 .../base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh     | 7 ++++++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/target/linux/ipq40xx/base-files/etc/board.d/02_network b/target/linux/ipq40xx/base-files/etc/board.d/02_network
index 7a4cac868cb78..f03df1b319cb1 100644
--- a/target/linux/ipq40xx/base-files/etc/board.d/02_network
+++ b/target/linux/ipq40xx/base-files/etc/board.d/02_network
@@ -152,6 +152,7 @@ ipq40xx_setup_macs()
 	linksys,mr8300)
 		wan_mac=$(mtd_get_mac_ascii devinfo hw_mac_addr)
 		lan_mac=$(macaddr_add "$wan_mac" 1)
+		label_mac=$wan_mac
 		;;
 	mikrotik,cap-ac |\
 	mikrotik,hap-ac2|\
diff --git a/target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh b/target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh
index ce0b55bb80915..4c4211cfc5b1f 100644
--- a/target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh
+++ b/target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh
@@ -26,8 +26,13 @@ preinit_set_mac_address() {
 	linksys,ea8300|\
 	linksys,mr8300)
 		base_mac=$(mtd_get_mac_ascii devinfo hw_mac_addr)
+		base_mac_la=$(macaddr_setbit "$base_mac" 7)
+		ip link set dev wan address "$base_mac"
 		ip link set dev lan1 address $(macaddr_add "$base_mac" 1)
-		ip link set dev eth0 address $(macaddr_setbit "$base_mac" 7)
+		ip link set dev lan2 address $(macaddr_add "$base_mac_la" 2)
+		ip link set dev lan3 address $(macaddr_add "$base_mac_la" 3)
+		ip link set dev lan4 address $(macaddr_add "$base_mac_la" 4)
+		ip link set dev eth0 address $(macaddr_add "$base_mac_la" 5)
 		;;
 	mikrotik,wap-ac|\
 	mikrotik,wap-ac-lte|\
